src/sds_data_model/_dataframe.py

```diff
@@ -283,64 +403,160 @@ def _create_dummy_dataset(
+    dataset = merge(
+        _create_empty_dataset(
+            column,
+            nodata[column],
+            lookup,
+            _metadata,
+            dtype[column],
+            dims,
+            height,
+            width,
+        )
+        for column in columns
+    )
+    dataset.update(
+        affine_to_coords(
+            transform,
+            height=height,
+            width=width,
+            y_dim=dims[0],
+            x_dim=dims[1],
+        ),
+    )
+    dataset.rio.write_crs("EPSG:27700", inplace=True)
+    dataset.rio.write_transform(transform, inplace=True)
+    dataset.attrs.clear()
+    dataset.to_zarr(
+        store=path,
+        mode="w",
+        compute=False,
+    )
+
+
+def _to_zarr_region(
+    pdf: PandasDataFrame,
+    path: str,
+    columns: List[str],
+    dtype: Dict[str, str],
+    nodata: Dict[str, float],
+    dataset_name: str,
+    cell_size: int = CELL_SIZE,
+    out_shape: Tuple[int, int] = OUT_SHAPE,
+    bng_ymax: int = BNG_YMAX,
+    geometry_column_name: str = "geometry",
+) -> PandasDataFrame:
+    """Rasterises a BNG grid cell and writes to the relevant region of a `zarr` file.
+
+    This function will be converted into a `pandas_udf`_ and called by `applyInPandas`_
+    within the :meth:`DataFrameWrapper.to_zarr` method.
+
+    Writing to a zarr file is a side effect of the method hence the input and the
+    output is the same Pandas DataFrame.
+
+    This function assumes that the Pandas DataFrame contains a column with BNG bounds
+    that is named "bounds".
+
+    Args:
+        pdf (PandasDataFrame): A Pandas DataFrame.
+        path (str): Path to save the zarr file including file name.
+        cell_size (int): The resolution of the cells in the DataArray. Defaults to
+            CELL_SIZE.
+        out_shape (Tuple[int, int]): The shape (height, width) of the DataArray.
+            Defaults to OUT_SHAPE.
+        bng_ymax (int): The maximum y value of the British National Grid.
+            Defaults to BNG_YMAX.
+        geometry_column_name (str): The name of the geometry column in the
+            Pandas DataFrame. Defaults to "geometry".
+
+    Returns:
+        PandasDataFrame: The input Pandas DataFrame is returned unchanged.
+
+    .. _`pandas_udf`:
+        https://spark.apache.org/docs/3.1.2/api/python/reference/api/pyspark.sql.functions.pandas_udf.html  # noqa: B950
+
+    .. _`applyInPandas`:
+        https://spark.apache.org/docs/3.1.2/api/python/reference/api/pyspark.sql.GroupedData.applyInPandas.htm
+    """
+    minx, miny, maxx, maxy = pdf["bounds"][0]
+    transform = Affine(cell_size, 0, minx, 0, -cell_size, maxy)
+    gpdf = GeoDataFrame(
+        data=pdf,
+        geometry=GeoSeries.from_wkb(pdf[geometry_column_name]),
+        crs="EPSG:27700",
+    )
+
+    if not columns:
+        columns = [dataset_name]
+        shapes = [gpdf.geometry]
+    else:
+        shapes = (zip(gpdf.geometry, gpdf[column]) for column in columns)
```

Comment:
You're getting a `mypy` error here because shapes could either be a `List[BaseGeometry]` or a `Generator[Iterable[Tuple[BaseGeometry, Any]], None, None]`. You can fix this by specifying the type before assignment in the if statement (assuming that you import all these types):

```suggestion
    shapes: Union[List[BaseGeometry], Generator[Iterable[Tuple[BaseGeometry, Any]], None, None]]
    if not columns:
        columns = [dataset_name]
        shapes = [gpdf.geometry]
    else:
        shapes = (zip(gpdf.geometry, gpdf[column]) for column in columns)
```